### PR TITLE
AP_RangeFinder: add support for HC-SR04 rangefinder

### PR DESCRIPTION
**Pull request by peterbarker**
Bench-tested only at this point using a PixRacer and PWM outputs 5 and 6 (pin 55 as echo pin and 54 as trigger pin).  Device was powered off the servo rail which was powered from an ESC.

Needs flight testing.

### CODE REVIEW

**Comment by Hwurzburg** [comment]
from @rmackay9  comment, I had assumed that it would consume too much cpu : https://github.com/ArduPilot/ardupilot_wiki/issues/789 and was working on a writeup of a Nano+US-015 (better version of SR04) for wiki, but this is much simpler, of course! will try to test on my copter this week

**Comment by peterbarker** [comment]
@Hwurzburg I wonder if the time spent parsing the serial streams for the arduino-based solution might exceed that of doing the measurement ourselves ;-)

File: libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    if (!hal.gpio->attach_interrupt(
+            params.pin,
+            FUNCTOR_BIND_MEMBER(&AP_RangeFinder_HC_SR04::irq_handler,
+                                void,
+                                uint8_t,
+                                bool,
+                                uint32_t),
+            AP_HAL::GPIO::INTERRUPT_BOTH)) {
+        // failed to attach interrupt
+        gcs().send_text(MAV_SEVERITY_WARNING,
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp:93]: I wonder if this is the best way to get the message to the user.. maybe the range finder should fail to initialise so it appears as a bad lidar health.

**Comment by rmackay9** [comment]
If it turns out the sonar doesn't work well then I think we should consider *not* adding support for it.  Adding support for a poor quality sensor may just generate more support calls.  Alternatively if we write up a wiki page put a big warning at the top that it's not recommend (if it doesn't work well).

**Comment by peterbarker** [comment]
On Sun, 17 May 2020, Randy Mackay wrote:

> +        // failed to attach interrupt
> +        gcs().send_text(MAV_SEVERITY_WARNING,
> 
> I wonder if this is the best way to get the message to the user.. maybe the range finder should fail to initialise so it appears as a bad lidar health.

This was swiped from AP_RSSI, so if we want to fix it we need to change it 
in both places.

This is also a runtime parameter, post-initialisation.  The thing should 
go unhealthy as there won't be readings.

**Comment by peterbarker** [comment]
On Sun, 17 May 2020, Randy Mackay wrote:

> If it turns out the sonar doesn't work well then I think we should consider not adding support for it. Adding support for a poor quality sensor may just
> generate more support calls. Alternatively if we write up a wiki page put a big warning at the top that it's not recommend (if it doesn't work well).

I'm looking forward to the flight test results, and consider @hwurzburg 
brave for volunteering :-)

Hopefully he has it running but not used for his initial flights....

I was actually wondering whether this would be useful as a flare detector 
in Plane or not.  I'm guessing "not", but worth a shot.

**Comment by khancyr** [comment]
I am a little unsure if we need to support those sensor, my main issue is they aren't reliable and you can have them with very very poor quality ... I really don't want to spend time to say on discuss that they got a crapy sensor and they should have invested in a better one ... like for APM boards...

**Comment by muramura** [comment]
I am PRing the ultrasonic sensor SRF02 at #7423.
I did a field test with the SRF02. There was a lot of variability.
It would be great if you could upload the actual measurement data as well as the manufacturer's published values.
I think that if the range of variability is large, there should be a filter or a correction that takes the average value.

File: libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    if (value_us == 0) {
+        // no reading; check for timeout:
+        if (now - last_reading_ms > 1000) {
+            // no reading for a second - something is broken
+            state.distance_cm = 0.0f;
+        }
+    } else {
+        // gcs().send_text(MAV_SEVERITY_WARNING, "Pong!");
+        // a new reading - convert time to distance
+        state.distance_cm = value_us / 58.0f;  // 58 is from datasheet
```
Comment (Hwurzburg) [inline libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp:159]: would a mult be better here in terms of cpu time?

**Comment by Hwurzburg** [comment]
@peterbarker I have is connected and working? before I try to fly with it a few inconsistencies:

the reported rangefinder1 value in MP is apparently the roundtrip uS, and the sonnarrange appears to be roughly 2x the range...I have been told  that the first should be cm to target while the sonnarrange should be meters to target....and while trying to find the answer, I saw that type 30 is not listed in the AP_Rangefinder.cpp file....being a code dummy I dont know if this is important or not..

File: libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp:159]: Yes!

**Comment by Hwurzburg** [comment]
generally its working pretty well....I set the max range at 2m, and this looks about right from the logs...above 2m it gets noisy...below 1.75m it holds alt over gully and embankments and long term drifting flight over level pretty well...if I had been using the baro, I would have had to reset the alt several times to prevent drifting into the ground....I put some GCS messages in to tell me when it switches from using RF to EKF...during the noise periods its rapidly switching .....no visible impact to flight...but if we did a .5s averager on it, I think not only would the alt source toggling quiet down, but higher operating height would be obtained....I have a US-015 sonar that is supposed to be less noise and more accurate, that I will try next AND I have a Maxbotix I2C EZ4 that actually can be made to work ECHO/TRIG instead of I2C....how hard would it be to try a .5s averager on the SRC-04 reports?

**Comment by Hwurzburg** [comment]
https://www.dropbox.com/s/eqc4lae6qlt0tsz/src04tesst.jpg?dl=0
https://www.dropbox.com/s/qt5gm47095tj4xk/src04test_1.jpg?dl=0

**Comment by peterbarker** [comment]
On Sun, 24 May 2020, Henry Wurzburg wrote:

> I have a Maxbotix I2C EZ4 that actually can be made to work ECHO/TRIG instead of I2C....how hard would it be to try a .5s averager on the SRC-04 reports?

I think we really try to avoid that sort of averaging in the drivers 
themselves.  There are definitely exceptions to that rule (uLanding), but 
generally we know that noise is What EKFs Crave.  Other places smooth the 
RangeFinder (and also eliminate noise!), so we ought not to over-smooth 
things :-)

We could, perhaps, say that >3m is out-of-range-high if things are bad 
enough?

Thanks very much for testing this out!  Really happy to hear that it 
worked for your test flights!

**Comment by rmackay9** [comment]
@peterbarker, I agree with the comment about overs smoothing in general but in general the EKF only makes very limited use of the range finder.  So for example it is required if the user also has an optical flow sensor.  The EKF's altitude is not affected by the rangefinder in general.  Terrain following for example does not involve the EKF.

**Comment by Hwurzburg** [comment]
@peterbarker I will test it over a smooth surface emulating a floor and see what alt it gets noisy and let you know for the out of range limit in the driver

**Comment by Hwurzburg** [comment]
[https://www.dropbox.com/s/sbuwp32zqaqhmtu/src04test_overflathard.jpg?dl=0](url)
test over concrete.....2m still the best operational range, maybe a 3m out of range in the code?

note, much better alt hold than EKF when I disable RF....interestingly CTUN.Alt does not seem better than raw baro...if estimated alt was baro instead of ekf estimate, it would have been more stable without the rangefinder operating....go figure? unless I miss-interpret the data...

**Comment by Hwurzburg** [comment]
https://www.dropbox.com/s/4xfppc43p1idgr3/US-015overhardsurface.jpg?dl=0
US-015 test....for and "improved" version this thing is really bad..very noisy....reflected in hold performance...it bounced around 0.25 meters with quick little alt moves...ekf hold was smoother, although would drift enough to require resets...maybe I got a dud....but I prefer the SRC-04 hands down...next the Maxbotix MB1242 in Echo/Trigger mode instead of I2C

**Comment by Hwurzburg** [comment]
tests with Maxbotix 1282 based sensor in Echo/Trig mode:
https://www.dropbox.com/s/v4vddni8k8rzdg1/1282raw.jpg?dl=0
notice never a negative glitch...and all miss-reads are WAY off...even with that it held the best so far of the tests..
I know we dont like to filte low level, but could not resist...very simple glitch detector...if measurement is more than 50cm either way from last one, ignore it...dont think it can get stuck unless there is a real failure...but a huge difference...this would allow its use much higher ,maybe up to its stated range of 7m...
https://www.dropbox.com/s/ikuzi8sltesvjb3/1282withglitchfltr.jpg?dl=0

I am going to see how much the glitch filter in the driver helps the SRC-04 next

**Comment by peterbarker** [comment]
@muramura you might like to see if your SRF02 suffers from the same glitching as Henry has found?

**Comment by Hwurzburg** [comment]
Final test results with the glitch filter:
Raw SRC-04
image
with glitch filter:
image

Raw Maxbotix1282 in echo/trigger mode instead of I2C:
image
with glitch filter:
image

recommend SRC-04 upper range as 2m and Maxbotix1282 as 4m (wiki as 7m now in I2C, will test to see if I2C mode is different than these results)
since this drive works for SRC-04, US-015 (need to test with glitch filter yet), and Maxbotix1282, perhaps its should be named an ECHO/TRGR driver instead of SRC-04 driver?

**Comment by peterbarker** [comment]
`AP_RangeFinder_TiggerAndPulseWidth` ?

**Comment by Hwurzburg** [comment]
SGTM

**Pull request merged**